lightningd/plugin_hook.c

```diff
@@ -434,31 +418,33 @@ static struct hook_node *find_hook(struct hook_node *graph, const char *name)
 	return NULL;
 }
 
-/* We maintain load order for ready-to-be-added hooks. */
-static void add_to_s(struct hook_node ***s, struct hook_node *n)
+/* Sometimes naive is best. */
+static struct hook_node *get_best_candidate(struct hook_node *graph)
 {
-	size_t pos, s_num;
+	struct hook_node *best = NULL;
 
-	s_num = tal_count(*s);
-	for (pos = 0; pos < s_num; pos++) {
-		if ((*s)[pos]->hook->plugin->index > n->hook->plugin->index)
-			break;
+	for (size_t i = 0; i < tal_count(graph); i++) {
+		if (graph[i].finished)
+			continue;
+		if (graph[i].num_incoming != 0)
+			continue;
+		if (!best
+		    || best->hook->plugin->index > graph[i].hook->plugin->index)
```

Comment:
We're never moving the plugin in the `graph`, so under no circumstance should this ever be true (a later iteration should always have a higher `index`). Since that's the case I think we can remove the `index` altogether, and break/return after finding a first `best`

Reply:
That's making an assumption that graph is in index order.  It (I'm fairly sure) starts that way, but can be perturbed by previous calls to plugin_hook_make_ordered.  If plugin B says it has to be before A and after C, we'll go C B A.  If B gets removed, we should return to A C.

Does anyone care?  Probably not, but this way is clear, at least.

Reply:
On further thought, if we define it as "minimum perturbation to achieve required ordering" which is a totally reasonable goal, we can indeed do this simplification.

Reply:
... Nope.  Hooks start in "who returned from getmanifest" order which is not the same as "order specified by config" order.